Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm

```diff
@@ -37,121 +37,205 @@
 
 namespace WebKit {
 
-NetworkTransportSession::NetworkTransportSession(NetworkConnectionToWebProcess& connection, nw_connection_group_t connectionGroup, nw_endpoint_t endpoint)
+NetworkTransportSession::NetworkTransportSession(NetworkConnectionToWebProcess& connection, nw_connection_t datagramConnection, nw_listener_t listener, nw_endpoint_t endpoint)
     : m_connectionToWebProcess(connection)
-    , m_connectionGroup(connectionGroup)
+    , m_listener(listener)
     , m_endpoint(endpoint)
 {
-    constexpr uint32_t maximumMessageSize { std::numeric_limits<uint32_t>::max() };
-    constexpr bool rejectOversizedMessages { false };
-    nw_connection_group_set_receive_handler(connectionGroup, maximumMessageSize, rejectOversizedMessages, makeBlockPtr([weakThis = WeakPtr { *this }] (dispatch_data_t datagram, nw_content_context_t, bool) {
-        RefPtr strongThis = weakThis.get();
-        if (!strongThis)
-            return;
+    m_connections.append(datagramConnection);
+    setupListenerConnectionHandler();
+    receiveDatagramLoop();
+}
 
-        // FIXME: Not only is this an unnecessary string copy, but it's also something that should probably be in WTF or FragmentedSharedBuffer.
-        auto vectorFromData = [](dispatch_data_t content) {
-            ASSERT(content);
-            Vector<uint8_t> request;
-            dispatch_data_apply_span(content, [&](std::span<const uint8_t> data) {
-                request.append(data);
-                return true;
-            });
-            return request;
-        };
+static RetainPtr<nw_parameters_t> createDatagramParameters()
+{
+    auto configureWebTransport = [](nw_protocol_options_t options) {
+        nw_webtransport_options_set_is_unidirectional(options, false);
+        nw_webtransport_options_set_is_datagram(options, true);
+    };
 
-        strongThis->receiveDatagram(vectorFromData(datagram).span());
-    }).get());
+    auto configureTLS = [](nw_protocol_options_t options) {
+        RetainPtr securityOptions = adoptNS(nw_tls_copy_sec_protocol_options(options));
+        sec_protocol_options_set_peer_authentication_required(securityOptions.get(), true);
+        sec_protocol_options_set_verify_block(securityOptions.get(), makeBlockPtr([](sec_protocol_metadata_t metadata, sec_trust_t trust, sec_protocol_verify_complete_t completion) {
+            // FIXME: Hook this up with WKNavigationDelegate.didReceiveChallenge.
+            completion(true);
+        }).get(), dispatch_get_main_queue());
+        // FIXME: Pipe client cert auth into this too, probably.
+    };
+
+    auto configureQUIC = [](nw_protocol_options_t options) {
+        nw_quic_set_initial_max_streams_bidirectional(options, 1000);
+        nw_quic_set_initial_max_streams_unidirectional(options, 1000);
+        nw_quic_set_max_datagram_frame_size(options, 65535);
```

Comment:
Is there a WebTransport spec that says to do this?  Can we use std::numeric_limits<uint16_t>::max() instead?